### PR TITLE
Add ioSSLSniffer

### PR DESCRIPTION
**Pull request by netanelc305**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# Prerequisites
+
+## Enable logging global
+
+This package relays on the ability to modify Apples logging global.   **requires a jailbroken device**
+
+### Howto
+
+- Download and install [`rpc_server`](https://github.com/doronz88/rpc-project) on a jailbroken device.
+- Set logging global to `True`
```
Comment (doronz88) [inline README.md:21]: Reformat making it clearer for the different ways

File: README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  - Using integrated `rpc_client`:
+    ```commandline
+    python3 -m iosslsniffer -o out.pcapng -ri RPC_IP -rp RPC_PORT
+    ```
+  - Using `rpc_client` shell
+
+## Apple's CFNetwork profile
+
+This profile is meant for debugging processes using the CFNetwork
```
Comment (doronz88) [inline README.md:30]: use RPC instead of the profile installation

File: iosslsniffer/sniffer.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class EntryHash:
+    pid: int
+    process_name: str
+    image: str
+    domain: str
+
+
+class Sniffer:
+    def __init__(self, lockdown: LockdownClient, out_file: str = 'tmp.pcapng', filters=None):
+        self._filters = filters
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:36]: init either your publics first

File: iosslsniffer/sniffer.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                continue
+
+            if 'done using Connection' in entry:
+                fd = int(entry.split()[-1])
+                if fd in self.active_fd:
+                    entry_hash, stream = self.active_fd[fd]
+                    logger.info(
+                        f'Saving recording for {entry_hash.process_name}({entry_hash.pid})  💾  {entry_hash.domain}')
+                    stream.dump(self._out_file)
+                    del self.active_fd[fd]
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:55]: in general, avoid using `del`. use `pop` instead

File: iosslsniffer/sniffer.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+            if 'SSL' not in entry:
+                continue
+            try:
+                parsed = self.syslog_parser.parse(entry)
+            except ParseFailedError:
+                continue
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:62]: should you produce a warning?

File: iosslsniffer/sniffer.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    def sniff(self) -> None:
+        for entry in self.os_trace_service.watch():
+            if 'CFNetwork' not in entry:
+                continue
+
+            if 'done using Connection' in entry:
+                fd = int(entry.split()[-1])
+                if fd in self.active_fd:
+                    entry_hash, stream = self.active_fd[fd]
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:51]: use `namedtuple` for accessing this dataset

File: iosslsniffer/sniffer.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            fd = parsed.fd
+            if fd not in self.active_fd:
+                logger.info(
+                    f'Recording SSL traffic {entry_hash.process_name}({parsed.pid}) ➡️  {entry_hash.domain}')
+                self.active_fd[fd] = (entry_hash, SSLSession(fd, parsed.local, parsed.peer))
+
+            self.active_fd[fd][1].process(parsed)
+
+    def should_keep(self, entry_hash: EntryHash) -> bool:
+        """Filter out entry if one of the criteria specified (pid,image,process_name)"""
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:81]: put spaces before and after the docstring prefix and postfix

File: iosslsniffer/sniffer.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        if self._filters.pids and entry_hash.pid in self._filters.pids:
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:82]: isn't this the opposite of what we want? if user included a pid list we should keep it if `entry_hash.pid in self._filters.pids`

File: iosslsniffer/sniffer.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            return False
+
+        if self._filters.images and entry_hash.image in self._filters.images:
+            return False
+
+        if self._filters.process_names and entry_hash.process_name in self._filters.process_names:
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:88]: these statemenets can be compiled into a single return statement

File: iosslsniffer/syslog_parser.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    """Parse syslog messages related to SSL"""
+    PATTERN = (
+        '{date} iPhone {filename}({image_name})[{pid:d}] <Notice>: CFNetwork Diagnostics [{}:{msg_id:d}] {} {\n'
+        '{ fd: {fd:d}, local {local:Local} => peer {peer:Peer} {domain}} {method} {extra}: (null)\n'
+        '{} data [ {data_size:d} ] bytes {\n{body:Body}}\n} [{}]')
+
+    HEXDUMP = r'\d{8}:\s?(.+?)(?=\s{4})'
+
+    @staticmethod
+    def parse(message: str) -> NamedTuple:
```
Comment (doronz88) [inline iosslsniffer/syslog_parser.py:20]: use a dedicated namedtuple type

File: iosslsniffer/tcp_simulation.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
+import logging
+import random
+from typing import Tuple
+
+# flake8: noqa
+logging.getLogger("scapy.runtime").disabled = True
```
Comment (doronz88) [inline iosslsniffer/tcp_simulation.py:6]: use `'`

File: iosslsniffer/tcp_simulation.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
+import logging
+import random
+from typing import Tuple
+
+# flake8: noqa
```
Comment (doronz88) [inline iosslsniffer/tcp_simulation.py:5]: use a specific `noqa`

File: iosslsniffer/tcp_simulation.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        """Send FIN packt"""
+        pkt = self._pkt('F')
+        self.seq_next += 1
+        return pkt
+
+    def _pkt(self, flags: str, ack: int = None) -> Packet:
+        """Packet builder"""
+        pkt = self.template.copy()
+
+        if ack:
```
Comment (doronz88) [inline iosslsniffer/tcp_simulation.py:52]: be more explicit: `if ack is not None`

File: iosslsniffer/tcp_simulation.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        pkt[TCP].seq = self.seq_next
+        pkt[TCP].flags = flags
+
+        return pkt
+
+
+class TCPSimSession:
+    """Simulate TCP Session between 2 peers"""
+    def __init__(self, src: Tuple[str, int], dst: Tuple[str, int]):
+        self.peer1 = TCPSimPeer(src, dst)
```
Comment (doronz88) [inline iosslsniffer/tcp_simulation.py:64]: rename the two peers into `src` and `dst`

File: README.md [hunk 15]
```diff
@@ -0,0 +1,61 @@
+# Description
+
+Simple pure python utility for sniffing HTTP/HTTPS decrypted traffic recorded by one of Apple's not-so-well documented
+APIs.
+
```
Comment (doronz88) [inline README.md:5]: - explain more about the workflow
- we sniff the RAW decrypted data, then simulate a "semi-equivalent" TCP stream to be able to view diagnose this with other tools...

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: iosslsniffer/sniffer.py (hunk 3 above)
Comment (netanelc305) [reply iosslsniffer/sniffer.py:36]: ?

File: README.md (hunk 2 above)
Comment (netanelc305) [reply README.md:30]: ?

File: iosslsniffer/__main__.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        return value
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.command('setup')
+@click.option('rpc', '-r', '--rpc', nargs=2, type=click.Tuple([str, int]), default=('127.0.0.1', 5910),
```
Comment (doronz88) [inline iosslsniffer/__main__.py:30]: import this port from the `rpcclient` module

File: iosslsniffer/__main__.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+              help='rpc server ip and port')
+def cli_setup(rpc):
+    """ Setup all prerequisites required inorder to sniff the SSL traffic """
+    rpc = create_tcp_client(rpc[0], rpc[1])
+    global_preferences = rpc.preferences.sc.open(GLOBAL_PREFERENCE)
+    if not global_preferences.get(CFNETWORK_KEY):
+        global_preferences.set(CFNETWORK_KEY, 3)
+        logger.info(f'{CFNETWORK_KEY} is now set, a restart is require')
```
Comment (doronz88) [inline iosslsniffer/__main__.py:38]: required

File: iosslsniffer/__main__.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        return
+    logger.info(f'{CFNETWORK_KEY} already set')
+    rpc.syslog.set_harlogger_for_all(True)
+
+
+@cli.command('sniff')
+@click.option('out_file', '-o', '--out', default='traffic.pcapng', type=click.STRING,
```
Comment (doronz88) [inline iosslsniffer/__main__.py:45]: use `click.Path()`

File: iosslsniffer/__main__.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+              help='outfile name with .pcapng extension')
+@click.option('pids', '-p', '--pid', type=click.INT, multiple=True, callback=default_none, help='filter pid list')
+@click.option('process_names', '-pn', '--process-name', callback=default_none, multiple=True,
+              help='filter process name list')
+@click.option('images', '-i', '--image', multiple=True, callback=default_none, help='filter image list')
+@click.option('--black-list/--white-list', default=True, is_flag=True)
+def cli_sniff(pids, process_names, images, out_file, black_list):
+    """ Sniff the traffic """
+    lockdown = create_using_usbmux()
```
Comment (doronz88) [inline iosslsniffer/__main__.py:54]: import `Command` from `pymobiledevice3` to have it select the device you work on and initialize `lockdown` for you

File: iosslsniffer/sniffer.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+
+
+@dataclass
+class FdEntry:
+    entry_hash: EntryHash
+    ssl_session: SSLSession
+
+
+class Sniffer:
+    def __init__(self, lockdown: LockdownClient, out_file: str = 'traffic.pcapng', filters=None):
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:50]: use default in const

File: iosslsniffer/sniffer.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+            except ParseFailedError:
+                continue
+
+            entry_hash = EntryHash(parsed.pid,
+                                   posixpath.basename(parsed.filename),
+                                   os.path.basename(parsed.image_name),
+                                   parsed.domain)
+
+            if not self._filters.should_keep(entry_hash):
+                logger.info(
```
Comment (doronz88) [inline iosslsniffer/sniffer.py:88]: change into a warning

File: iosslsniffer/tcp_simulation.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+logging.getLogger('scapy.runtime').disabled = True
+
+from scapy.layers.inet import IP, TCP
+from scapy.layers.l2 import Ether
+from scapy.packet import Packet, Raw
+from scapy.utils import wrpcap
+
+
+class TCPPeer:
+    """Simulate a TCP Peer"""
```
Comment (doronz88) [inline iosslsniffer/tcp_simulation.py:15]: spaces

**Review by doronz88 — CHANGES_REQUESTED** [review]

**Pull request merged**